% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/SClineager_io.R
\name{read_sclineager}
\alias{read_sclineager}
\title{Read datsets for SClineager}
\usage{
read_sclineager(
  runinfo,
  coverage_cutoff,
  coverage_percentage,
  cell_percentage,
  folder,
  artefact_percentage,
  species = "hg38",
  keep_other = FALSE
)
}
\arguments{
\item{runinfo}{A data frame with ``Path'' and ``Cell''. ``Path'' is the path to the raw mutation output folder for each cell. ``Cell'' is the name of each single cell.}

\item{coverage_cutoff}{Coverage below this cutoff is considered not observable.}

\item{coverage_percentage}{A numeric value in [0,1]. If a mutation has more than \code{coverage_percentage} of cells whose coverages are below \code{coverage_cutoff}, then it will not considered in modeling.}

\item{cell_percentage}{A numeric value in [0,1]. If a cell has more than \code{cell_percentage} of mutations not observable, then it will not considered in modeling.}

\item{folder}{A folder at which outputs are saved.}

\item{artefact_percentage}{A numeric value in [0,1]. Only keep mutations that appeared in more than \code{artefact_percentage} of cells.}

\item{species}{A character value that indicates germline_mutations. Default is ``hg38'' and other possibilities are ``hg19'' and ``mm10''.}

\item{keep_other}{A logical value whether to keep intronic and intergenic mutations. Default is \code{FALSE}.}
}
\value{
``summary.pdf'' and ``results.RData'' are saved under \code{folder}. The RData file contains a list with components (also returned):
\item{mutations_mat}{The variant allele frequency (VAF) matrix, whose rows are mutations and columns are samples.}
\item{coverage_mat}{The total sequencing coverage matrix, whose rows are mutations and columns are samples.}
\item{runinfo}{Same as \code{runinfo} in the input, but trimmed to contain only cells that appeared in \code{mutations_mat}, in the same order.}
\item{annotation}{Annotation information of the variants in \code{mutations_mat}.}
}
\description{
Read datsets for SClineager
}
\examples{
# See the github page https://github.com/inmybrain/SClineager.
}
